draft-ietf-jsonpath-base.md

```diff
@@ -1195,19 +1188,159 @@ compared is an object, array, boolean, or `null`.
 
-The semantics of regular expressions are as defined in {{-iregexp}}.
+##### Function Extensions {#fnex}
+{: unnumbered}
 
-##### Boolean Operators
+[^unnumbered-bad]
+
+[^unnumbered-bad]: This should not be an unnumbered section, which it is
+    forced to be by the parent section being unnumbered too.
+    The overall structure where the parts of this subsection go to
+    needs to be decided; one part already is in the IANA
+    Considerations in {{iana-fnex}}.
+    This PR will focus on content, a future PR will fix the structure.
+{:source=" -- cabo"}
+
+Beyond the filter expression functionality defined in the preceding
+subsections, JSONPath defines an extension point that can be used to
+add filter expression functionality: "Function Extensions".
+
+A function extension defines a registered name (see {{iana-fnex}}) that
+can be applied to a sequence of zero or more arguments, resulting in a
+result.
+A function-expression stands for ("returns") a nodelist or a value.
+The arguments are filter expressions that are to be interpreted as
+nodelists or values.
+For each argument and for the result, the function extension defines
+the "kind" of the item, i.e., whether the item is a value or a
+nodelist ("nodes").
+
+~~~ abnf
+function-name           = function-name-first *function-name-char
+function-name-first     = LCALPHA
+function-name-char      = DIGIT / function-name-first / "_"
+LCALPHA                 = %x61-7A  ; "a".."z"
+
+
+function-expression     = function-name "(" S [function-argument
+                             *(S "," S function-argument)] S ")"
+singular-path           =/ function-expression
+function-argument       = comparable
+~~~
+
+Syntactically, a function-expression can occur anywhere where a
+singular-path can occur (exist-expr, comparable).
+A function-expression that employs a function extension that returns a
+nodelist MUST return a singular node to be used in a comparable.
+
+###### Function Extensions: length {#length}
 {: unnumbered}
 
-The logical AND, OR, and NOT operators have the normal semantics of Boolean algebra and
-obey its laws (see, for example, {{BOOLEAN-LAWS}}).
+Arguments:
+: 1. value
+
+Result:
+: value (unsigned integer)
+
+The "length" function extension provides a way to compute the length
+of a value and make that available for further processing in the
+filter expression:
+
+~~~ JSONPath
+$[?length(@.authors) >= 5]
+~~~
+
+Its only argument is a value (possibly taken from a singular path as
+in the example above).  The result also is a value, an unsigned
+integer.
+
+* If the argument value is a string, the result is the number of
+  Unicode scalar values in the string.
+* If the argument value is an array, the result is the number of
+  elements in the array.
+* If the argument value is an object, the result is the number of
+  members in the object.
+* For any other argument value, the result is one.
+
+
+###### Function Extensions: count {#count}
```

Comment:
What's the difference between `count()` and `length()`?  Why have a distinction?

Reply:
``` 
###### Function Extensions: length {#length}
{: unnumbered}

Arguments:
: 1. value

###### Function Extensions: count {#count}
{: unnumbered}

Arguments:
: 1. nodelist
```

One is for JSON values, one is for nodelists.

Reply:
This isn't apparent to me from the text.  Some additional clarification would be good. (May be provided with the examples I asked for in another comment.)